docs/docs/tooling/profiler.md

```diff
@@ -0,0 +1,127 @@
+---
+title: Noir Profiler
+description: Learn about the Noir Profiler, how to generate execution flamegraphs, identify bottlenecks, and visualize optimizations. 
+keywords: [profiling, profiler, flamegraph] 
+sidebar_position: 0
+---
+
+## Noir Profiler
+
+`noir-profiler` is a sampling profiler designed to analyze and visualize Noir programs. It assists developers to identify bottlenecks by mapping execution data back to the original source code.  
+
+### Installation
+
+`noir-profiler` comes out of the box with [noirup](../getting_started/noir_installation.md). Test that you have the profiler installed by running `noir-profiler --version`. 
+
+### Usage
+
+Let's start by creating a simple Noir program. All this program aims to do is zero out an array past some dynamic index.
+
+```rust
+fn main(ptr: pub u32, mut array: [u32; 32]) -> pub [u32; 32] {
+    for i in 0..32 {
+        if i > ptr {
+            array[i] = 0;
+        }
+    }
+    array
+}
+```
+You can use these values for the `Prover.toml`:
+```toml
+ptr = 1
+array = [1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1]
+```
+
+Running `nargo info` we can get some information about the opcodes produced by this program, but it doesn't give us a lot of info on its own. Compile and execute this program normally using `nargo compile` and `nargo execute`. 
+
+### Generating an ACIR opcode flamegraph
+
+The program on its own is quite high-level. Let's get a more granular look at what is happening by using `noir-profiler`. 
+
+After compiling the program, run the following:
+```sh
+noir-profiler opcodes --artifact-path ./target/program.json -output ./target/
+```
+Below you can see an example flamegraph with a total 387 opcodes (using `nargo` version 1.0.0-beta.2):
+<picture>
+  <img src="../../static/img/tooling/profiler/acir-flamegraph-unoptimized.png">
+</picture>
+
+You should now have a flamegraph mapping ACIR opcodes to their associated locations in the source code. We highly recommend generating these graphs yourself while going through this guide. You will be able to click into call stacks and view the flamegraph much more clearly. 
```

Comment:
I think it's worth mentioning here that opening it in a browser allows you to inspect more closely due to being able to click into various regions of the flamegraph whereas image viewers generally don't.

Reply:
Updating this comment to mention the interactive nature of the browser

Reply:
Awesome, thanks.